File: arduino/Stepper-motors/src/helpers/path/line.cpp
#include <math.h>

#include "line.h"

Line::Line(double xStart, double yStart, double xEnd, double yEnd,  double maxSpeed, double maxAcceleration, bool precision):Path()
{
    mPrecision = precision;

    mStart[X] = xStart;
    mStart[Y] = yStart;

    mEnd[X] = xEnd;
    mEnd[Y] = yEnd;

    mMaxSpeed = maxSpeed;
    mMaxAcceleration =  maxAcceleration;

    if(mMaxSpeed < 0) {
        mMaxSpeed = -mMaxSpeed;
        mGoingBackwards = true;
    }

    double distance = sqrt(
        pow(mStart[X]-mEnd[X], 2) + 
        pow(mStart[Y]-mEnd[Y], 2)
    );

    double accelerationTime = mMaxSpeed / mMaxAcceleration;
    double accelerationDistance =  0.5 * pow(accelerationTime, 2) * mMaxAcceleration;

    if(distance == 0)
    {
        mVector[X]=1;
        mVector[Y]=0;
        mTheta = 0;
        mExpectedDuration = 0;
    }
    else
    {
        mVector[X] = (mEnd[X]-mStart[X]) / distance;
        mVector[Y] = (mEnd[Y]-mStart[Y]) / distance;

        mTheta = acos(mVector[X]);

        if(mVector[Y]<0)
            mTheta = -mTheta.toDouble()+2* M_PI;

        if(2 * accelerationDistance > distance)
        {
            // Short distance, we don't have time to reach full speed
            mDistanceForAcceleration = distance / 2;
            mDistanceWithConstantSpeed = 0;
            accelerationTime = sqrt(2 * mDistanceForAcceleration / mMaxAcceleration);

            mEndAccelerationTime = accelerationTime * SEC_TO_MICROS_MULTIPLICATOR;
            mStartDeccelerationTime = mEndAccelerationTime + 0;
            mExpectedDuration = 2 * mEndAccelerationTime;
        }
        else
        {
            // Long distance, we will reach full speed
            mDistanceForAcceleration = accelerationDistance;
            mDistanceWithConstantSpeed = distance - 2 * accelerationDistance;

            mEndAccelerationTime =  accelerationTime * SEC_TO_MICROS_MULTIPLICATOR;
            mStartDeccelerationTime = mEndAccelerationTime + mDistanceWithConstantSpeed / mMaxSpeed * SEC_TO_MICROS_MULTIPLICATOR;
            mExpectedDuration = mStartDeccelerationTime + accelerationTime * SEC_TO_MICROS_MULTIPLICATOR;
        }
  }
}

Line::~Line(){}

String Line::debugString() 
{
  String xStart = String(mStart[X], 1);
  String yStart = String(mStart[Y], 1);
  String xEnd = String(mEnd[X], 1);
  String yEnd = String(mEnd[Y], 1);
  return String("Line from (" + xStart + "," + yStart + ") to (" + xEnd + "," + yEnd + ")");
}

double Line::positionError(double x, double y, Angle theta, long time)
{
  long ellapsedTime = time - mStartTime;

  // We use a base where vector X is colinear to the path
  double projectedX = (x-mStart[X])*mVector[X]+(y-mStart[Y])*mVector[Y];

  if(ellapsedTime < 0)
  {
    double vectorX = cos(theta.toDouble());
    double vectorY = sin(theta.toDouble());
    projectedX = (mStart[X]-x) * vectorX + (mStart[Y]-y) * vectorY;
    return projectedX;
  }
  else if(0 < ellapsedTime && ellapsedTime < mEndAccelerationTime)
  {
    double x = 0.5*mMaxAcceleration*pow((ellapsedTime) * MICROS_TO_SEC_MULTIPLICATOR, 2);
    return x-projectedX;
  }
  else if(mEndAccelerationTime < ellapsedTime && ellapsedTime < mStartDeccelerationTime)
  {
    double x = mDistanceForAcceleration + (ellapsedTime- mEndAccelerationTime)* MICROS_TO_SEC_MULTIPLICATOR * mMaxSpeed;
    return x-projectedX;
  }
  else if(mStartDeccelerationTime < ellapsedTime && ellapsedTime < mExpectedDuration)
  {
    double x = 2 * mDistanceForAcceleration
        + mDistanceWithConstantSpeed 
        - 0.5 * mMaxAcceleration *pow((mExpectedDuration-ellapsedTime)* MICROS_TO_SEC_MULTIPLICATOR, 2);

    return x-projectedX;
  }
  else
  {
    double vectorX = cos(theta.toDouble());
    double vectorY = sin(theta.toDouble());
    projectedX = (mEnd[X]-x)*vectorX+(mEnd[Y]-y)*vectorY;
    return projectedX;
  }
}

double Line::rotationError(double x, double y, Angle theta, long time)
{
  // Conversion from perpendicular error to an angle
  const double ANGLE_ERROR_COEFFICIENT = M_PI_4 / (mMaxSpeed * 30);
  long ellapsedTime = time - mStartTime;

  double baseRotationError = (mTheta-theta).toDouble();

  // We project on the path line
  double projectedY = -(x-mStart[X])*mVector[Y]+(y-mStart[Y])*mVector[X];
  //return baseRotationError;

  if(ellapsedTime < 0)
  {
    return baseRotationError;
  }
  else if(0 < ellapsedTime && ellapsedTime < mEndAccelerationTime)
  {
    return baseRotationError - projectedY * ANGLE_ERROR_COEFFICIENT * ellapsedTime * MICROS_TO_SEC_MULTIPLICATOR * mMaxAcceleration;
  }
  else if(mEndAccelerationTime<ellapsedTime && ellapsedTime<mStartDeccelerationTime)
  {
    return baseRotationError - projectedY * ANGLE_ERROR_COEFFICIENT * mMaxSpeed;
  }
  else if(mStartDeccelerationTime<ellapsedTime && ellapsedTime<mExpectedDuration)
  {
    return baseRotationError - projectedY * ANGLE_ERROR_COEFFICIENT* MICROS_TO_SEC_MULTIPLICATOR*(mExpectedDuration-ellapsedTime)*mMaxAcceleration;
  }
  else
  {
    return baseRotationError;
  }
}

bool Line::needsPrecision() const {
  return mPrecision;
}

File: potato/src/path/line.py
from math import sqrt, pow, acos, cos, sin, pi
from src.location import Coordinates
from src.path.angle import Angle
from src.path.path import Path

class Line(Path):
    def __init__(self, x_start: float, y_start: float, x_end: float, y_end: float, going_backwards: bool, max_speed: float = 500, max_acceleration: float = 500) -> None:
        super().__init__(going_backwards)
        self._x_start = x_start
        self._y_start = y_start
        self._x_end = x_end
        self._y_end = y_end
        self._max_speed = max_speed
        self._max_acceleration = max_acceleration

        self._vector_x = 0.0
        self._vector_y = 0.0
        self._theta = Angle(0)

        self._acceleration_distance = 0.0
        self._distance_with_constant_speed = 0.0

        distance = sqrt(
            pow(self._x_start - self._x_end, 2) +
            pow(self._y_start - self._y_end, 2)
        )

        acceleration_time = self._max_speed / self._max_acceleration
        acceleration_distance = 0.5 * pow(acceleration_time, 2) * self._max_acceleration

        if distance == 0:
            self._vector_x = 1
            self._vector_y = 0
            self._theta = Angle(0)
        else:
            self._vector_x = (self._x_end - self._x_start) / distance
            self._vector_y = (self._y_end - self._y_start) / distance
            self._theta = Angle(acos(self._vector_x))

            if self._vector_y < 0:
                self._theta += (2 * pi)

            if self.is_going_backwards():
                self._theta += pi


            if 2 * acceleration_distance > distance:
                # Short distance, we don't have time to reach full speed
                self._acceleration_distance = distance / 2
                self._distance_with_constant_speed = 0
                acceleration_time = sqrt(2 * self._acceleration_distance / self._max_acceleration)

                self._end_acceleration_time = acceleration_time
                self._start_decceleration_time = acceleration_time
                self._expected_duration = 2 * acceleration_time
            else:
                # Long distance, we will reach full speed
                self._acceleration_distance = acceleration_distance
                self._distance_with_constant_speed = distance - 2 * acceleration_distance

                self._end_acceleration_time = acceleration_time
                self._start_decceleration_time = acceleration_time + self._distance_with_constant_speed / self._max_speed
                self._expected_duration = self._start_decceleration_time + acceleration_time

    def expected_position(self, time: float) -> Coordinates:
        ellapsed_time = time - self._start_time
        
        if ellapsed_time < 0:
            return Coordinates(self._x_start, self._y_start, self._theta)
        elif 0 < ellapsed_time < self._end_acceleration_time:
            expected_x = self._x_start + self._vector_x * 0.5 * self._max_acceleration * pow(ellapsed_time, 2)
            expected_y = self._y_start + self._vector_y * 0.5 * self._max_acceleration * pow(ellapsed_time, 2)
            return Coordinates(expected_x, expected_y, self._theta)
        elif self._end_acceleration_time < ellapsed_time < self._start_decceleration_time:
            expected_x = self._x_start + self._vector_x * (self._acceleration_distance + (ellapsed_time - self._end_acceleration_time) * self._max_speed)
            expected_y = self._y_start + self._vector_y * (self._acceleration_distance + (ellapsed_time - self._end_acceleration_time) * self._max_speed)
            return Coordinates(expected_x, expected_y, self._theta)
        elif self._start_decceleration_time < ellapsed_time < self._expected_duration:
            expected_x = self._x_end - self._vector_x * 0.5 * self._max_acceleration * pow(self._expected_duration - ellapsed_time, 2)
            expected_y = self._y_end - self._vector_y * 0.5 * self._max_acceleration * pow(self._expected_duration - ellapsed_time, 2)
            return Coordinates(expected_x, expected_y, self._theta)
        else:
            return Coordinates(self._x_end, self._y_end, self._theta)
